### PR TITLE
Pushover device support

### PR DESCRIPTION
**Pull request by alec-pinson**
## Description

Pushover have a parameter `device` which allows you to send notifications to a specific device rather than all devices.
This PR is to allow that as a valid option.

Related PR - https://github.com/prometheus/alertmanager/pull/3318

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note
Adds support for `device` parameter on pushover configs in alertmanager
```

### CODE REVIEW

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go [hunk 1]
```diff
@@ -715,6 +715,9 @@ type PushoverConfig struct {
 	URLTitle string `json:"urlTitle,omitempty"`
 	// The name of one of the sounds supported by device clients to override the user's default sound choice
```
Comment (JoaoBraveCoding) [inline pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:716]: Comment has to be updated

**Review by JoaoBraveCoding — COMMENTED** [review]
Small comment, but apart from that lgtm however this should wait for the AM PR to be merged

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go (hunk 1 above)
Comment (alec-pinson) [reply pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:716]: well spotted!

**Comment by alec-pinson** [comment]
> Small comment, but apart from that lgtm however this should wait for the AM PR to be merged

No probs, not sure how long they normally take to review/merge but I'll monitor it.
For now I'm using my own [alert manager image](https://github.com/users/alec-pinson/packages/container/package/alertmanager) and having to manually overwrite the `kube-prometheus-alertmanager-generated` secret.

**Comment by alec-pinson** [comment]
Hi @JoaoBraveCoding just thought I would let you know that the PR has just been merged

**Review by JoaoBraveCoding — COMMENTED** [review]
From my side lgtm but I don't have the right to merge

**Comment by simonpasquier** [comment]
yes ideally we should wait for the next Alertmanager version.

**Comment by alec-pinson** [comment]
The new version has just been released including my PR 👍
https://github.com/prometheus/alertmanager/releases/tag/v0.26.0

**Review by ArthurSens — COMMENTED** [review]
I believe you'll need to add an entry for Device here:

https://github.com/prometheus-operator/prometheus-operator/blob/34bb905e3ca90963ffdaa50cbe1924137f9a5ed7/pkg/alertmanager/amcfg.go#L1108-L1116

**Review by simonpasquier — CHANGES_REQUESTED** [review]
There's more work needed since the AlertmanagerConfig fields aren't propagated to the final configuration + we need some version checks.

see https://github.com/prometheus-operator/prometheus-operator/pull/4697 for an example.

**Comment by alec-pinson** [comment]
> There's more work needed since the AlertmanagerConfig fields aren't propagated to the final configuration + we need some version checks.
> 
> see #4697 for an example.

I don't currently have a development machine but as soon as I do I will try get this corrected 👍

**Comment by alec-pinson** [comment]
HI @simonpasquier , I've updated (hopefully) as requested. However the E2E tests failed, apparently alertmanager restarted 5 times... but I'm not sure why?

**Comment by ArthurSens** [comment]
Hey @alec-pinson, I've just tried running the failed e2e-test locally and I could see it pass. I've retriggered the CI here, it looks like flakiness 🤔

File: pkg/alertmanager/amcfg.go [hunk 2]
```diff
@@ -1126,6 +1126,7 @@ func (cb *configBuilder) convertPushoverConfig(ctx context.Context, in monitorin
 		URLTitle:      in.URLTitle,
 		Priority:      in.Priority,
 		HTML:          in.HTML,
+		Device:        *in.Device,
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:1129]: you need to check for nil

Comment (alec-pinson) [reply pkg/alertmanager/amcfg.go:1129]: Not sure what you mean? Sorry I'm not a real go dev 😁

Comment (simonpasquier) [reply pkg/alertmanager/amcfg.go:1129]: No problem, sorry for being dense!
```suggestion
		}
		if in.Device != nil {
			out.Device = *in.Device
		}
```

**Comment by alec-pinson** [comment]
wheyy its green

**Review by ArthurSens — APPROVED** [review]

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**